### PR TITLE
[Eager] Support 3 test case of test_imperative_basic in eager mode

### PR DESCRIPTION
**Pull request by wanghuancoder**
### PR types

New features
### PR changes

Others 
### Describe

In Eager mode, the 13 unit tests in test_imperative_basic have been tuned. The added logic includes:
1. The kernel is called, and the output tensor is passed into the original pointer of the tensor on the python side, so that the tensor on the python side can be modified
2. Add logic to handle the automatic call timing of RetainGradForTensor
3. _varbase_creator, create_parameter, support the creation of EagerTensor in Eager mode, including EagerParamBase class
4. Dynamic graph, Parameter Initializer no longer uses append_op, but uses _C_ops.XXX
5. Support Initialzier FP16 tests in eager mode
6. Support all dygraph_utils in eager mode

### CODE REVIEW

File: paddle/fluid/eager/auto_code_generator/eager_generator.cc [hunk 1]
```diff
@@ -1201,19 +1201,19 @@ static std::pair<std::string, std::string> GenerateForwardFunctionContents(
       // in form of shared_ptr<EagerTensor>/vector<shared_ptr<EagerTensor>>
       if (output.duplicable()) {
         const char* FWD_NUM_ARG_TEMPLATE =
-            ", std::vector<egr::EagerTensor>& %s";
+            ", std::vector<egr::EagerTensor*>& %s";
```
Comment (JiabinYang) [inline paddle/fluid/eager/auto_code_generator/eager_generator.cc:1204]: change code sample before to show what's to be created

File: paddle/fluid/eager/grad_node_info.cc [hunk 2]
```diff
@@ -47,20 +48,34 @@ void GradNodeBase::AddEdges(const std::vector<AutogradMeta*>& metas,
...
+    } else {
+      meta->SetGradNode(std::make_shared<egr::GradNodeAccumulation>());
+      adj_edges_[slot_id].emplace_back(meta->GetMutableGradNode(),
+                                       meta->OutRankInfo());
+    }
   }
 }
 
-void GradNodeBase::AddEdges(const AutogradMeta& meta, size_t slot_id) {
+void GradNodeBase::AddEdges(AutogradMeta& meta, size_t slot_id) {
```
Comment (JiabinYang) [inline paddle/fluid/eager/grad_node_info.cc]: Change this to pointer but not reference without const

File: paddle/fluid/eager/grad_node_info.h [hunk 3]
```diff
@@ -106,7 +106,7 @@ class GradNodeBase {
    * This one is called slot by slot
    * **/
   void AddEdges(const std::vector<AutogradMeta*>& metas, size_t slot_id);
-  void AddEdges(const AutogradMeta& meta, size_t slot_id);
+  void AddEdges(AutogradMeta& meta, size_t slot_id);  // NOLINT
```
Comment (JiabinYang) [inline paddle/fluid/eager/grad_node_info.h]: change this, NOLINT here is not acceptable

File: paddle/fluid/eager/utils.cc [hunk 4]
```diff
@@ -135,12 +135,14 @@ std::vector<std::shared_ptr<egr::EagerTensor>> EagerUtils::SyncToVars(
   return res;
 }
 
+void deleter(egr::EagerTensor* ptr) {}
```
Comment (JiabinYang) [inline paddle/fluid/eager/utils.cc]: change this to lamda

File: paddle/fluid/eager/utils.cc [hunk 5]
```diff
@@ -135,12 +135,14 @@ std::vector<std::shared_ptr<egr::EagerTensor>> EagerUtils::SyncToVars(
...
+
 static std::shared_ptr<egr::EagerTensor> TrySyncToVar(
     egr::EagerTensor* tensor) {
   if (tensor->initialized() || tensor->Var().IsInitialized()) {
     tensor->SyncToVar(paddle::framework::proto::VarType_Type_LOD_TENSOR);
   }
-  return std::make_shared<EagerTensor>(*tensor);
+  return std::shared_ptr<egr::EagerTensor>(tensor, deleter);
```
Comment (JiabinYang) [inline paddle/fluid/eager/utils.cc]: maybe lamda here is enough

File: paddle/fluid/pybind/eager.cc [hunk 6]
```diff
@@ -72,6 +73,15 @@ void EmptyEagerTensorInitializer(
...
                                   paddle::framework::make_ddim(dims)));
     self->eager_tensor.set_impl(dense_tensor);
+  } else {
+    PADDLE_THROW(platform::errors::InvalidArgument(
+        "We only support LoDTensor to be constructed by this initializer, "
+        "please check your var type first and make sure you are going to "
+        "construct LoDTensor."));
+  }
+
+  if (!autograd_meta->GetMutableGradNode()) {
```
Comment (JiabinYang) [inline paddle/fluid/pybind/eager.cc:83]: Add VLOG here since it's a vulnerable action

Comment (JiabinYang) [inline paddle/fluid/pybind/eager.cc:83]: Add VLOG here since it's a vulnerable action

File: paddle/fluid/pybind/eager_functions.cc [hunk 7]
```diff
@@ -115,7 +115,8 @@ static PyObject* eager_api_scale(PyObject* self, PyObject* args,
 static PyObject* eager_api_numpy_to_tensor(PyObject* numpy_data,
                                            pten::DataType dtype,
                                            const paddle::platform::Place& place,
-                                           bool stop_gradient) {
+                                           bool stop_gradient,
+                                           bool retain_grads) {
```
Comment (JiabinYang) [inline paddle/fluid/pybind/eager_functions.cc:119]: maybe we don't need this, using `autograd.to_tensor` is enough

File: paddle/fluid/pybind/eager_functions.cc [hunk 8]
```diff
@@ -203,6 +213,27 @@ static PyObject* eager_api_run_backward(PyObject* self, PyObject* args,
...
+      reinterpret_cast<EagerTensorObject*>(PyTuple_GET_ITEM(args, 0))
+          ->eager_tensor;
+  egr::EagerTensor& dst =
+      reinterpret_cast<EagerTensorObject*>(PyTuple_GET_ITEM(args, 1))
+          ->eager_tensor;
+  auto place = CastPyArg2Place(PyTuple_GET_ITEM(args, 2), 2);
+  bool blocking = CastPyArg2AttrBoolean(PyTuple_GET_ITEM(args, 3), 3);
+
+  dst = src.copy_to(pten::TransToPtenBackend(place), blocking);
+  egr::EagerUtils::autograd_meta(&dst)->SetStopGradient(true);
```
Comment (JiabinYang) [inline paddle/fluid/pybind/eager_functions.cc]: This should be as same as src tensor

File: paddle/fluid/pybind/eager_functions.cc [hunk 9]
```diff
@@ -219,6 +250,9 @@ PyMethodDef variable_functions[] = {
      METH_VARARGS | METH_KEYWORDS, NULL},
     {"run_backward", (PyCFunction)(void (*)(void))eager_api_run_backward,
      METH_VARARGS | METH_KEYWORDS, NULL},
+    {"eagertensor_copy",
```
Comment (JiabinYang) [inline paddle/fluid/pybind/eager_functions.cc]: change name to `tensor_copy`

File: paddle/fluid/pybind/eager_functions.cc [hunk 10]
```diff
@@ -203,6 +213,27 @@ static PyObject* eager_api_run_backward(PyObject* self, PyObject* args,
   EAGER_CATCH_AND_THROW_RETURN_NULL
 }
 
+static PyObject* eager_api_eagertensor_copy(PyObject* self, PyObject* args,
```
Comment (JiabinYang) [inline paddle/fluid/pybind/eager_functions.cc]: Change this to `eager_api_tensor_copy`

File: paddle/fluid/pybind/eager_properties.cc [hunk 11]
```diff
@@ -129,6 +145,31 @@ PyObject* eager_tensor_properties_get_dtype(EagerTensorObject* self,
   EAGER_CATCH_AND_THROW_RETURN_NULL
 }
 
+PyObject* eager_tensor_properties_get_retain_grads(EagerTensorObject* self,
```
Comment (JiabinYang) [inline paddle/fluid/pybind/eager_properties.cc:148]: This part of code still have something to discuss

File: python/paddle/fluid/dygraph/parallel.py [hunk 12]
```diff
@@ -322,7 +322,7 @@ def scale_loss(loss):
     if not ParallelEnv().world_size > 1:
         return loss
 
-    loss_scale = to_variable(
+    loss_scale = to_tensor(
```
Comment (JiabinYang) [inline python/paddle/fluid/dygraph/parallel.py:325]: why this?

**Review by JiabinYang — COMMENTED** [review]
some comments

File: paddle/fluid/eager/auto_code_generator/eager_generator.cc (hunk 1 above)
Comment (wanghuancoder) [reply paddle/fluid/eager/auto_code_generator/eager_generator.cc:1204]: done, thx!

File: paddle/fluid/eager/grad_node_info.cc (hunk 2 above)
Comment (wanghuancoder) [reply paddle/fluid/eager/grad_node_info.cc]: done, thx!

File: paddle/fluid/eager/grad_node_info.h (hunk 3 above)
Comment (wanghuancoder) [reply paddle/fluid/eager/grad_node_info.h]: done, thx!

File: paddle/fluid/eager/utils.cc (hunk 4 above)
Comment (wanghuancoder) [reply paddle/fluid/eager/utils.cc]: done, thx!

File: paddle/fluid/eager/utils.cc (hunk 5 above)
Comment (wanghuancoder) [reply paddle/fluid/eager/utils.cc]: done, thx!

File: paddle/fluid/pybind/eager.cc (hunk 6 above)
Comment (wanghuancoder) [reply paddle/fluid/pybind/eager.cc:83]: done, thx!

Comment (wanghuancoder) [reply paddle/fluid/pybind/eager.cc:83]: done, thx!

File: paddle/fluid/pybind/eager_functions.cc (hunk 8 above)
Comment (wanghuancoder) [reply paddle/fluid/pybind/eager_functions.cc]: done, thx!

File: paddle/fluid/pybind/eager_functions.cc (hunk 9 above)
Comment (wanghuancoder) [reply paddle/fluid/pybind/eager_functions.cc]: done, thx!

File: paddle/fluid/pybind/eager_functions.cc (hunk 10 above)
Comment (wanghuancoder) [reply paddle/fluid/pybind/eager_functions.cc]: done, thx!

File: python/paddle/fluid/dygraph/parallel.py (hunk 12 above)
Comment (wanghuancoder) [reply python/paddle/fluid/dygraph/parallel.py:325]: 因为scale_loss这个函数在eager_tensor_patch_methods.py中被用到了，目前Eager模式不支持to_variable

File: paddle/fluid/eager/utils.cc [hunk 13]
```diff
@@ -236,4 +252,18 @@ std::vector<EagerTensor> EagerUtils::RecoverTensorWrapper(
   return ret;
 }
 
+void EagerUtils::CheckRetainGrad(const egr::EagerTensor& tensor) {
+  if (FLAGS_retain_grad_for_all_tensor) {
+    egr::egr_utils_api::RetainGradForTensor(tensor);
+  }
+}
+
+void EagerUtils::CheckRetainGrad(const std::vector<egr::EagerTensor>& tensors) {
```
Comment (JiabinYang) [inline paddle/fluid/eager/utils.cc]: Bad name, `CheckAndRetainGrad`

File: python/paddle/fluid/dygraph/parallel.py (hunk 12 above)
Comment (JiabinYang) [reply python/paddle/fluid/dygraph/parallel.py:325]: Ok, i will support this later

File: python/paddle/tensor/creation.py [hunk 14]
```diff
@@ -121,7 +126,8 @@ def to_tensor(data, dtype=None, place=None, stop_gradient=True):
         if dtype is None:
             dtype = paddle.get_default_dtype()
         return core.eager.to_tensor(data,
```
Comment (JiabinYang) [inline python/paddle/tensor/creation.py:128]: actually this make cause problem since value maybe not `numpy.ndarray`

**Review by JiabinYang — COMMENTED** [review]
some comments

File: paddle/fluid/eager/utils.cc (hunk 13 above)
Comment (wanghuancoder) [reply paddle/fluid/eager/utils.cc]: thx, done!

File: python/paddle/tensor/creation.py (hunk 14 above)
Comment (wanghuancoder) [reply python/paddle/tensor/creation.py:128]: 目前，这个在eager_function.cc里有判断，只支持numpy。其它的情况会PADDLE_THROW，报错。

**Review by JiabinYang — APPROVED** [review]
LGTM

**Pull request closed without merging**